# CMSC320 Programming Assignment 3
Please go through the file following the instructions as you go.

Rename this notebook to `CMSC320_PA3_YourName.ipynb`

This is adapted from an assignment from Dr. Fardina Alam's section.

**NOTE:** It's by design that this assignment has more places for you to write code from scratch and potentially look up functions online. For example, we covered joins conceptually but didn't practice them much. The pandas code for joins is easy to look up online and adapt to your context.


---
---
**DISCLAIMER: This is a fake story, not a real incident! We have created this story just for the data exploration purpose.**

---
---

**Campus Police Department**

Incident Report

---

**Incident Date:** September 10, 2025

**Location of Incident:** 2nd Floor, Iribe Building

**Items Left at the Scene:** Red Ticket for Zichao's Magic Show

**Case Number:** 2025-IRB-0027

---
**Victim Information:**

**Name:** Dr. Fardina Alam

**Affiliation:** University of Maryland, College Park

**Position:** Professor

---

**Incident Description:**

On the evening of September 10, 2025, an unknown individual(s) unlawfully entered Dr. Fardina's office located on the 2nd floor of Iribe and stole a set of markers. The stolen items were a set of limited edition markers that Dr. Fardina had spent years collecting. A red entry ticket for "Zichao's Magic Show" was found on the floor near Dr. Fardina's desk.







In [1]:
# Run this cell
import pandas as pd

## Instructions:

In this assignment, you are tasked with solving the mystery of Dr. Fardina's stolen markers and finding the culprit.

To accomplish this, you will be provided with **4 data sets** containing key information related to the incident. The filenames are:

 - Profiles
 - Witnesses
 - Zichao's magic show
 - Interrogation Statements



Your goal is to analyze these datasets to uncover the clues and evidence that will lead you to the thief.

Take a moment to look at the dataset documentation below:

---

**Profiles:**

This dataset contains information regarding all the potential people that could have committed the crime. The columns are as follows:
- **ID:** A **unique** identifier for each person
- **Name:** The name of the person
- **Location_during_crime:** The location of the person during the crime
- **Eye_Color:** The eye color of the person
- **Skin_Color:** The skin color of the person
- **Hair_Color:** The hair color of the person
- **Height:** The height of the person in cm

**Witnesses:**

This dataset contains a statement from everyone in **Profiles** about suspicious things they saw during the time of the crime. The columns are as follows:
- **ID:** A **unique** identifier for each person  (this ID matches the **Profiles** **ID** column).
- **Statement:** A statement from the person regarding a suspicious activity during the time of the crime (text!).

**Zichao's magic show:**

This dataset contains everyone that purchased a ticket for the show. **NOTE** that not everyone in **Profiles** has purchased a ticket for this show. The columns are as follows:
- **ID:** A **unique** identifier for each person (this ID matches the **Profiles** **ID** column).
- **Ticket_Number:** A **unique** number identifying a specific sale.

**Interrogation Statements:**

This dataset would give you the statement from someone you wish to interrogate. You will zero in on one suspect over the course of the assignment and look up that person's statement in this table. Assume that each statement is 100% true. The columns are as follows:

- **Name:** The name of the person being interrogated (this name matches the **Profiles** **Name** column).
- **Statement:** A statement the person gives after being interrogated (text!)

---

## 1. Let's begin by looking at the Profiles dataset


**1.1 (1 point):**  In the code box below, load the "Profiles" dataset into a variable called **profiles_df** and display it. (Recall how to upload a file to Colab from PA2! Feel free to create additional cells but do make in-cell comments about what each new cell does.)


In [2]:
# Write your code here

profiles_df = pd.read_csv("Profiles.csv")
profiles_df

,ID,Name,Location_during_crime,Eye_Color,Skin_Color,Hair_Color,Height
0,P299,James Mason,fourth floor,red,indigo,orange,191
1,P086,Christopher Rodriguez,second floor,indigo,indigo,green,198
2,P022,Misty Ashley,fourth floor,blue,red,violet,154
3,P003,Amanda Brown,lobby,green,blue,violet,158
4,P036,William Jones,fourth floor,blue,yellow,orange,184
...,...,...,...,...,...,...,...
343,P126,Kenneth Watson,NaN,violet,indigo,red,176
344,P203,Donald Anderson,third floor,yellow,indigo,orange,193
345,P273,Harold Watkins,fourth floor,blue,indigo,red,193
346,P290,Brandon Bennett,third floor,NaN,yellow,indigo,161


**1.2 (1 point):**  Next, display the number of rows and columns in **profiles_df**.

In [3]:
# Write your code here

profiles_df.shape

# (rows, columns)

(348, 7)

**1.3 (1 point):**  Display the names of the columns in **profiles_df**.

In [4]:
# Write your code here

profiles_df.columns

Index(['ID', 'Name', 'Location_during_crime', 'Eye_Color', 'Skin_Color',
       'Hair_Color', 'Height'],
      dtype='object')

**1.4 (1 point):**  Now, display the datatypes of the columns in **profiles_df**.

In [5]:
# Write your code here

profiles_df.dtypes

ID                       object
Name                     object
Location_during_crime    object
Eye_Color                object
Skin_Color               object
Hair_Color               object
Height                    int64
dtype: object

**1.5 (1 point):** Use DATAFRAME_NAME.count() on the dataframe and pass the result to a new variable called **info**. Display **info**. This shows you how many non-NA values each column has. Write down any **potential problems** with the dataset that you see when examining the output of the count function. Hint: you may find it helpful to read the documentation for pandas' .count().




In [6]:
# Write your code here

info = profiles_df.count()
info

ID                       348
Name                     348
Location_during_crime    280
Eye_Color                280
Skin_Color               348
Hair_Color               348
Height                   348
dtype: int64

*Write your observations here:*

There appears to be two columns location_during_crime and eye_color that have fewer values than every other column, meaning that there will be potential missing values in the dataframe as we process it. 



## 2. Data cleaning

**2.1 (10 points):** We want to be able to use the rows with no provided eye color. What we'll do is fill those missing values with something.

***Please read the directions carefully:***


- In order to fill in a missing eye color for someone, we want to find the **most common eye color** among people who have the **same hair and skin color** as the person.

- For instance, if we want to fill in the eye color of someone who has yellow hair and violet skin, we replace the missing eye color with the eye color that we identified to be the most common in people with yellow hair and violet skin.

**HINT:** you may find it useful to first find the most common eye color for each combination of hair and skin colors so you don't do that more than once for each combination of hair and skin features.

**NOTE:** in case you find multiple modes for some group, choose the color that appears first in alphabetical order.

When you finish filling in the missing eye colors, please display **profiles_df**.


In [7]:
# Write your code here
# Remember to feel free to add code cells as needed

profiles_df['Eye_Color'] = profiles_df.groupby(
['Hair_Color','Skin_Color'])['Eye_Color'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else x))

# defining and using (transforming) lambda function to fill in most common eye colors for each combination of hair and skin colors

profiles_df

,ID,Name,Location_during_crime,Eye_Color,Skin_Color,Hair_Color,Height
0,P299,James Mason,fourth floor,red,indigo,orange,191
1,P086,Christopher Rodriguez,second floor,indigo,indigo,green,198
2,P022,Misty Ashley,fourth floor,blue,red,violet,154
3,P003,Amanda Brown,lobby,green,blue,violet,158
4,P036,William Jones,fourth floor,blue,yellow,orange,184
...,...,...,...,...,...,...,...
343,P126,Kenneth Watson,NaN,violet,indigo,red,176
344,P203,Donald Anderson,third floor,yellow,indigo,orange,193
345,P273,Harold Watkins,fourth floor,blue,indigo,red,193
346,P290,Brandon Bennett,third floor,yellow,yellow,indigo,161



## 3. Some data summaries

Now that we have successfully cleaned our **Profiles** data, we can summarize and visualize some of it.

**3.1 (2 points):** Do a DATAFRAME_NAME.describe() and reflect on what you see.

In [8]:
# Write your code here

profiles_df.describe()

,Height
count,348.000000
mean,175.425287
std,15.539951
min,150.000000
25%,161.000000
50%,175.000000
75%,190.000000
max,200.000000


*Reflect here:*

The only numerical column in the entire dataset is height as it was the only column that returned with descriptive statistics, central tendency, and spread calculations for the dataframe's distribution. However, within this singular column there is a rather wide range of 348 heights that average out at around 175 and max out at 200 (High standard deviation). The other 6 columns are categorical in nature. 

**3.2 (2 points):** Learn what a Pandas cross-tab is and make a cross-tab table of eye color vs hair color.

In [9]:
# Write your code here

pd.crosstab(profiles_df['Eye_Color'], profiles_df['Hair_Color'])

Hair_Color,blue,green,indigo,orange,red,violet,yellow
Eye_Color,,,,,,,
blue,1,8,1,7,8,8,6
green,19,11,2,5,10,11,4
indigo,8,6,8,10,3,1,1
orange,7,9,7,3,14,9,10
red,12,5,6,6,9,3,13
violet,4,5,4,8,7,1,7
yellow,9,1,15,11,6,11,8


**3.3 (2 points):** Which columns are categorical and which are numerical?

*Answer here:*

Numerical: Height

Categorical: Eye_Color, Hair_Color, ID, Skin_Color, Name, and Locating_during_crime.


## 4. Figuring out the features of the culprit

Now that we have successfully cleaned our **Profiles** data and learned a bit more about it, we notice that all the other datasets are actually clean. Let's begin figuring out the features of the criminal from witness testimonies. Take a moment to look at the **"Witnesses"** table. We won't necessarily want all the statements from every single witness.

**NOTE:** Take a look at the structure of each statement; notice how there are only a few variations of statements from the witnesses.

**4.1 (4 points):** Find all witnesses that were near the crime scene
- First load in the Witness.csv file and pass it to a variable called "**witness_df**"
- create a new dataframe called **"important_witnesses"** that contains the id and the witness statements of those who were near the scene of the crime
- display the new dataframe when you are finished

**HINT:** look at the floor the the crime happened on.

In [10]:
# Write your code here

# crime happened on the second floor

witness_df = pd.read_csv("Witness.csv")
display(witness_df)

merged_df = witness_df.merge(profiles_df, on='ID') # inner join to get only ID matches 

important_witnesses = merged_df[merged_df['Location_during_crime'].str.strip().str.lower() == 'second floor'][['ID', 'Statement']].drop_duplicates(subset='ID') # normalized/standardized potential data inconsistencies (drops duplicates from one-to-many relationship --> multiple rows for the same IDs/statements in profile table)

important_witnesses



,ID,Statement
0,P001,I saw someone with red hair acting suspicious.
1,P002,I saw someone with violet hair acting suspicious.
2,P003,I saw someone who looked 170 cm tall acting su...
3,P004,I saw someone with yellow eyes acting suspicious.
4,P005,I saw someone with yellow hair acting suspicious.
...,...,...
231,P295,I saw someone with blue skin acting suspicious.
232,P296,I saw someone with blue skin acting suspicious.
233,P297,I saw someone with violet hair acting suspicious.
234,P298,I saw someone with red eyes acting suspicious.


,ID,Statement
6,P007,I saw someone with red skin acting suspicious.
8,P009,I saw someone who looked 150 cm tall acting su...
11,P013,I saw someone with red skin acting suspicious.
12,P015,I saw someone with orange hair acting suspicious.
19,P024,I saw someone who looked 170 cm tall acting su...
26,P031,I saw someone with red skin acting suspicious.
27,P033,I saw someone who looked 164 cm tall acting su...
29,P035,I saw someone with orange hair acting suspicious.
38,P045,I saw someone who looked 165 cm tall acting su...
46,P055,I saw someone who looked 155 cm tall acting su...


In [11]:
important_witnesses.count()

ID           48
Statement    48
dtype: int64

**4.2 (10 points):** Based on the important witnesses' testimonies, find every potential eye color that the culprit could have.

**Meaning,** find a list of eye colors mentioned by the important witnesses.

 - Make a table called **"potential_eye_colors"** which should have potential eye colors and how many times each of them was mentioned in someone's statement. Display it when done.

 **NOTE:** You shouldn't be collating this data by hand. You CAN use For loops, but it's also possible to do this without For loops.

In [12]:
# Write your code here

potential_eye_colors = important_witnesses['Statement'].str.extract(r'with (\w+) eyes')[0].value_counts()
potential_eye_colors

# regex to extract the first word following "with" and preceding "eyes" and return it as one column (series)

0
orange    5
blue      2
violet    1
Name: count, dtype: int64

**4.3 (4 points):** Find every potential hair color that the culprit could have

 - Make a table called **"potential_hair_colors"** which should have potential hair colors and how many times each of them was mentioned in someone's statement. Display it when done.

In [13]:
# Write your code here

potential_hair_colors = important_witnesses['Statement'].str.extract(r'with (\w+) hair')[0].value_counts()
potential_hair_colors

# regex to extract the first word following "with" and preceding "hair" and return it as one column (series)

0
green     6
orange    5
violet    3
blue      2
Name: count, dtype: int64

**4.4 (4 points):** Find every potential skin color that the culprit could have

 - Make a table called **"potential_skin_colors"** which should have potential skin colors and how many times each of them was mentioned in someone's statement. Display it when done.

In [14]:
# Write your code here

potential_skin_colors = important_witnesses['Statement'].str.extract(r'with (\w+) skin')[0].value_counts()
potential_skin_colors

# regex to extract the first word following "with" and preceding "skin", and return it as one column (series)

0
red       6
orange    3
green     1
Name: count, dtype: int64

**4.5 (10 points):** Find the potential height of the culprit

 - Put the smallest and largest potential heights of the culprit in the variables called **"smallest_height" & "largest_height"** and display them when you are done

In [15]:
# Write your code here

heights = important_witnesses['Statement'].str.extract(r'(\d+) cm')[0].dropna().astype(int)

# regex to extract height digits 0-9 preceding cm string and populate one column with the results as integers (series, dropped rows that don't mention numeric height before converting to int so that I could use min/max functions properly)

smallest_height = heights.min()
largest_height = heights.max()

print("Smallest Height:", smallest_height) 
print("Largest Height:", largest_height)

Smallest Height: 150
Largest Height: 170


## 5. Finding who might own the ticket

**5.1 (1 point):** Load **"Zichao_s Magic Show.csv"** into a dataframe called **"show_df"** and display it when you are done.

In [16]:
# Write your code here

show_df = pd.read_csv('Zichao_s Magic Show.csv')
show_df

,ID,Ticket_Number
0,P132,50269975
1,P056,28120732
2,P250,88051289
3,P197,79545169
4,P012,83521845
...,...,...
146,P121,31341288
147,P003,85008158
148,P180,24491860
149,P283,38524355



**5.2 (8 points):**  Remember that there was a show ticket left at the crime scene, and it'd be helpful to know which tickets match the color of the ticket at the scene.

Create a column called **Ticket_Color** for the **"show_df"** that displays the colors of each ticket based on its Ticket ID.

How do you find the ticket color from the ticket number to populate the new column? **You asked the ticket issuer and learned the rule**: Sum the individual digits in a ticket number, then look at the rightmost digit and use this mapping:

0-2: Blue, 3-5: Red, 6-7: Green, 8-9: Yellow.

Display the **"show_df"** when you are done.

In [17]:
# Write your code here

def ticket_color(ticket):
    digit_sum = sum(int(d) for d in str(ticket)) # looping through each digit as a string before converting to integer and summing them up
    rightmost = digit_sum % 10 # right most digit following summing of the ticket number (modulus remainder will always be the right digit)
    
    if rightmost <= 2: # 0-2
        return "Blue"
    elif rightmost <= 5: # 3-5
        return "Red"
    elif rightmost <= 7: # 6-7
        return "Green"
    else:
        return "Yellow" # 8-9

show_df["Ticket_Color"] = show_df["Ticket_Number"].apply(ticket_color) # applying function defined to each ticket number and storing result in new ticket color column
show_df

# function to specify ticket color rule

# confirming that the ticket number is converted to integer

# summing integers/each digit up

,ID,Ticket_Number,Ticket_Color
0,P132,50269975,Red
1,P056,28120732,Red
2,P250,88051289,Blue
3,P197,79545169,Green
4,P012,83521845,Green
...,...,...,...
146,P121,31341288,Blue
147,P003,85008158,Red
148,P180,24491860,Red
149,P283,38524355,Red


**5.2 (3 points):** Add the ticket color to the **Profiles** dataframe. It should retain all the **Profiles** rows, and include a ticket color for the row if that person has a ticket. Display **Profiles**.

In [18]:
profiles_df = profiles_df.merge(show_df[['ID','Ticket_Color']], on='ID', how='left')

profiles_df # left join to get every row from profiles table and only matches from ticket show table (if the person has a ticket the color is display, if not the color is 'NaN')

,ID,Name,Location_during_crime,Eye_Color,Skin_Color,Hair_Color,Height,Ticket_Color
0,P299,James Mason,fourth floor,red,indigo,orange,191,NaN
1,P086,Christopher Rodriguez,second floor,indigo,indigo,green,198,NaN
2,P022,Misty Ashley,fourth floor,blue,red,violet,154,NaN
3,P003,Amanda Brown,lobby,green,blue,violet,158,Red
4,P036,William Jones,fourth floor,blue,yellow,orange,184,NaN
...,...,...,...,...,...,...,...,...
343,P126,Kenneth Watson,NaN,violet,indigo,red,176,NaN
344,P203,Donald Anderson,third floor,yellow,indigo,orange,193,Green
345,P273,Harold Watkins,fourth floor,blue,indigo,red,193,Yellow
346,P290,Brandon Bennett,third floor,yellow,yellow,indigo,161,NaN


## 6. Finding the culprit

A helpful function might be the "iloc" function. Have a look at: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html

**6.1 (10 points):** Using all the data compiled from before find a person in **Profiles** that matches all the features you've discovered. Put the name of the culprit in a variable called **"culprit"**

**NOTE:** Each feature (eye color, hair color, skin color, height) need to have been mentioned by some witness. For example, the eye color doesn't have to be the most common value in a list of mentioned eye colors; it just needs to be present.

**NOTE 2:** The culprit should have a red ticket.

Describe your thought process in the text cell.

In [19]:
# Write your code here

suspects = profiles_df[
    (profiles_df['Height'].between(150,170)) &
    (profiles_df['Location_during_crime'] == "second floor") &
    (profiles_df['Ticket_Color'] == "Red") &
    (profiles_df['Skin_Color'].isin(["green", "orange", "red"])) &
    (profiles_df['Hair_Color'].isin(["green", "orange", "violet", "blue"])) &
    (profiles_df['Eye_Color'].isin(["orange", "violet", "blue"]))
]

# contains mentioned eye, hair, and skin colors
# contains height range
# contains ticket color as red
# contains location as second floor

display(suspects)

culprit = "Ishan Kar"

,ID,Name,Location_during_crime,Eye_Color,Skin_Color,Hair_Color,Height,Ticket_Color
283,P202,Ishan Kar,second floor,orange,green,blue,170,Red


*Describe your thought process here:*

**6.2 (4 points):** Check if your culprit is correct!

Once you've identified one name, you can interrogate them. What they would say can be located next to their name in **"Interrogation_Statements.csv"**.

Load the dataset into **"interrogation_df"** and print the statement associated with the name you've identified below.

In [20]:
# Write your code here

interrogation_df = pd.read_csv("Interrogation_Statements.csv")

interrogation_df[interrogation_df['Name'] == culprit]

,Name,Statement
468,Ishan Kar,Fardina's Markers were just too dang cool :( I...


## 7. Feedback

**7.1 (1 point):** Please share what you enjoyed and didn't enjoy about this assignment and how long it took for you to do it.

*Share your thoughts here:*

I think this assignment was interactive enough for me to learn new things about python and the way dataframes interact with one another in my loaded environment. I enjoyed having to decipher the scenario of the given data through pandas and applying that within 4 different datasets to filter out and single the result to one person. However, I didn't enjoy how long this assignment ended up taking me because I truly wanted to explore every aspect that it was asking us to answer. From searching up the individual methods to understanding what each of them did exactly in the context of the question I was being asked to complete. It was a bit of a handful in this regard. 

## 8. Submitting your work

### Congrats on finishing PA3!

1. Save this notebook (`File > Save`).
2. Download it to your computer as `.ipynb` (`File > Download > Download .ipynb`).
3. Submit the .ipynb file and the new .csv file on Canvas.

Tips:
1. After finishing the work, click on the arrow to the right of "Run all" and then click on "Clear all outputs" and "Restart session." Then run each cell in order to make sure everything compiles.
2. Remember to delete the cells you created for testing and any Gemini-affected cells if you used help from Gemini.